Create scs-XXXX-vN-security-groups-decision-record.md (#495)
SovereignCloudStack/standards: Standards/scs-XXXX-vN-security-groups-decision-record.md

---
title: Security Groups Decision Record
type: Decision Record
status: Draft
track: IaaS
author: josephineSei
date: DD-03-2024
---

## Introduction

Security Groups in IaaS (OpenStack) are sets of ip table rules, that are applied to ports which connect a VM to a network.

> In contrast to other resources like flavors or volume types that are always publicly accessable, or images that can be both publically and private, security groups are always bound to the project level.

Reviewer: ```suggestion
In contrast to other resources like flavors or volume types that are always publicly accessable, or images that can be both public and private, security groups are always bound to the project level.
```

## Terminology

Security Group
  A set of iptables rules that is applied to ports connecting a VM and a network.

Security Group Rule (abbr. Rule)
  This references a single rule within a security group.

RBAC
  Role Based Access Control used for policies and alike.

network rbac / neutron rbac
  These access controls will let administrators and users share neutron related resources with other projects.

admin
  The most powerful role in OpenStack. Only given to some Operators of the Cloud Infratructure.

## Context

While creating a VM and also later on, one or more security groups can be added to it.
When there is no security group specified the default security group will always be added.
Like every other security group, the default group is also project bound.
That means, it can be edited as required by project members.

When thinking about standardizing security groups, the goal is to automatically have more than one default security group.
But to also have predefined groups with predefined roles for certain and often used cases as e.g. ssh, http and https (maybe also icmp).

> Those groups should best be available to all projects and predefined by administrators.

Reviewer 2: I edited the above comment for better readability, I didn't change the meaning, hopefully.
Some sentences where confusing to read, so I hopefully reworded them to be easier to understand.

> Those groups should best be available to all projects and pre-defined by administrators.

Reviewer: ```suggestion
Those groups should best be available to all projects and predefined by administrators.
```

As security groups are project bound and there is no native way to them to be shared, we are left with two options:

1. To use another endpoint `network rbac` to share security groups among different projects.
2. To adhere to the project scope of security groups and only give documentation about recommended security groups to users.

### Option 1: operator usage of network rbac

> The `network rbac` endpoint[^1] manages the possibitity to share and access certain network sepcific resources as security groups.

Reviewer 3: ```suggestion
The `network rbac` endpoint[^1] manages the possibitity to share and access certain network-specific resources such as security groups.
```

For admins it is possible to use this endpoint to share a security group with ALL projects within the the cloud including ALL projects of ALL domains:

```bash
stack@devstack:~/devstack$ openstack network rbac create --target-all-projects --action access_as_shared --type security_group group-for-everyone
+-------------------+--------------------------------------+
| Field             | Value                                |
+-------------------+--------------------------------------+
| action            | access_as_shared                     |
| id                | 394d8e71-143f-4c5b-a72f-cd10af3222df |
| object_id         | b6a3834a-f89c-47a9-9ed6-ca89e93701c4 |
| object_type       | security_group                       |
| project_id        | 15f2ab0eaa5b4372b759bde609e86224     |
| target_project_id | *                                    |
+-------------------+--------------------------------------+
```

> This would fulfill our goal to grant access to pre-defined security groups fro all projects and all groups recieved as shared do not count into the projects quota for security groups.

Reviewer: ```suggestion
This would fulfill our goal to grant access to predefined security groups fro all projects and all groups recieved as shared do not count into the projects quota for security groups.
```

stack@devstack:~/devstack$ openstack network rbac show bc22a865-46f9-4cd2-80af-3c249ba0e010
+-------------------+--------------------------------------+
| Field             | Value                                |
+-------------------+--------------------------------------+
| action            | access_as_shared                     |
| id                | bc22a865-46f9-4cd2-80af-3c249ba0e010 |
| object_id         | 5f22e42e-95dc-4c0a-8651-57b1dfc8639f |
| object_type       | security_group                       |
| project_id        | 15f2ab0eaa5b4372b759bde609e86224     |
| target_project_id | *                                    |
+-------------------+--------------------------------------+
```

The biggest downside: As soon as a security group is shared, everyone from every project, can edit the rules of this group.

[^1]: [Neutron RBAC](https://docs.openstack.org/neutron/latest/admin/config-rbac.html)

### Option 2: stay project-scoped

Using and adhering the project scope of the security groups has the consequence, that:

1. either an admin has to set up security groups for each project

> 2. or the scs only provides a guide on how to setup and use some recommended security groups.

Reviewer: ```suggestion
2. or the SCS project only provides a guide on how to setup and use some recommended security groups.
```

> 2. or the scs only provides a guide on how to setup and use some recommended security groups.

Reviewer: Whom is that guide aimed for? Normal user? Reseller?, ...

Author: I already wrote a draft for such a guide. We should aim it for users, because they (may) need the guidance: https://github.com/SovereignCloudStack/docs/pull/142

> 

Reviewer 2: there is always the option of doing none of this, which is Option 3.

Option 4:
We might add a tutorial on how to create your own security group in general, how to detect what kind of network permissions your project needs for most frequent (linux) workloads and how to craft your own security groups in a **secure way**.

We might carefully add one or two examples that are really reviewed for security, but I'm not a fan of providing examples, see my long response above as to why.

As users are allowed to, will and should edit their security groups, there is no way to ensure, that a certain set of security groups with certain rules is always present in a project.
So packing an extra burden on admins is unreasonable.
The option to give a guide and recommend a few security groups however is a quite good option.

## Decision

> To be decided.

Reviewer: We need a decision here. I prefer option 2, writing a guide.

Reviewer 3: I also prefer option 2 here since shared security groups are still mutable at runtime affecting all consuming projects and resources immediately if any changes to them are applied. This sounds problematic from various angles.

Author: I wrote a first version of the decision, as everyone is for Option 2 maybe mixed with Option 3

## Consequences

> What becomes easier or more difficult to do because of this change?

Author: I will add the consequences after we have a decision.

---
title: Security Groups Decision Record
type: Decision Record
status: Draft
track: IaaS
author: josephineSei
date: DD-03-2024
---

## Introduction

Security Groups in IaaS (OpenStack) are sets of ip table rules, that are applied to ports which connect a VM to a network.
In contrast to other resources like flavors or volume types that are always publicly accessible, or images that can be both public and private, security groups are always bound to the project level.
That creates some difficulties for a possible standard of Security Groups, which are discussed in this document.

## Terminology

Security Group

>   A set of iptables rules that is applied to ports connecting a VM and a network.

Reviewer: ```suggestion
  A set of iptables rules that is applied to ports connecting a Virtual Machine and a network.
```

Security Group Rule (abbr. Rule)
  This references a single rule within a security group.

RBAC
  Role Based Access Control used for policies and alike.

network rbac / neutron rbac
  These access controls will let administrators and users share neutron related resources with other projects.

admin
  The most powerful role in OpenStack. Only given to some Operators of the Cloud Infrastructure.

## Context

While creating a VM and also later on, one or more security groups can be added to it.
When there is no security group specified the default security group will always be added.
Like every other security group, the default group is also project bound.
That means, it can be edited as required by project members.

> By design of OpenStack and when not changed the default rules in the default security group block ALL incoming traffic and only allow outgoing traffic[^1].

Reviewer: ```suggestion
By design of OpenStack and when not changed, default rules in default security group block ALL incoming traffic and only allow outgoing traffic[^1].
```

Reviewer 4: Is that true? I think the default traditionally allowed for incoming traffic originating from the SAME group.

Reviewer 4: If we want to diverge from the traditional behavior, we need to communicate this will with our users.
An upstream discussion to align this between OpenStack providers even beyond SCS would also make sense.

Reviewer 2: I haven't yet found time to look into the code, but what @garloff wrote during the call today seems to indicate that he's right.

Should we schedule another out of band meeting to discuss this, how do we want to handle this?

Author: I was relying on the Security Groups document from the Nova team: https://docs.openstack.org/nova/latest/user/security-groups.html (also mentioned in the DR). The second paragraph in there is: "All projects have a default security group which is applied to any port that has no other security group defined. Unless you change the default, this security group denies all incoming traffic and allows only outgoing traffic from your instance."

So maybe we do have an inconsistency in the Upstream Nova docs too.
But as this DR will only decide, whether there should be a set of pre-defined security groups or just a guide. This discussion belongs into the issue that will standardize the default security group rules: https://github.com/SovereignCloudStack/standards/issues/521 

Mittlerweile (seit bobcat) ist es tatsächlich default, dass auch ingress von einer bestimmten SG erlaubt ist:
https://github.com/openstack/neutron/blob/master/neutron/db/migration/alembic_migrations/versions/2023.2/expand/c33da356b165_security_group_default_rules.py

Ich werde diesen Abschnitt des DR umschreiben und im anderen Issue genauer darauf eingehen.
Es wäre auch gut zu wissen, wann (ab welchem release) sich der default von gar kein ingress zu ingress nur von derselben SG geändert hat.

[^1]: [Security Group Overview](https://docs.openstack.org/nova/latest/user/security-groups.html)

### Reasons for and against a standard for security groups

Considering having most likely similiar security groups within different projects, it might make sense to standardize a few security groups for often used cases like ssh, http, https and maybe icmp.
What speaks for standardizing a certain set of security groups:

1. Having a set of correctly configured security groups could reduce misconfiguration from users
2. Re-using correctly configured security groups saves time for users
3. Auditing security groups would be way easier for operators when helping customers

> 4. The configuration for the SGs can be done by networking experts, which may result in a higher security level as when users without expertise configure them

Reviewer: ```suggestion
4. The configuration for the security groups can be done by networking experts, which may result in a higher security level as when users without expertise configure them
```

| 92a79600-5358-4fef-    | None        | IPv4      | 0.0.0.0/0 |            | egress    | None                  | None                 | True                           | True                          |
| a390-822665f46070      |             |           |           |            |           |                       |                      |                                |                               |
| 997bb0c2-652e-4d1f-    | None        | IPv4      | 0.0.0.0/0 |            | ingress   | PARENT                | None                 | True                           | False                         |
| b910-e12c89f88b44      |             |           |           |            |           |                       |                      |                                |                               |
+------------------------+-------------+-----------+-----------+------------+-----------+-----------------------+----------------------+--------------------------------+-------------------------------+
```

Those rules can be edited, which may pose a security risk for customers consuming the default security group.

> This should be adressed as an pre-requirement [here](https://github.com/SovereignCloudStack/standards/issues/521).

Reviewer: ```suggestion
This should be adressed as a pre-requirement [here](https://github.com/SovereignCloudStack/standards/issues/521).
```
